android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMManager.java

```diff
@@ -719,6 +720,66 @@ public boolean accept(File pathname) {
     }
   }
 
+  /**
+   * Migrate keyboards list from the legacy dat file (HashMap) and returns a list of keyboards.
+   * 1. These deprecated keyboards: "us", "european", "european2" are also removed
+   * and replaced with "sil_euro_latin".
+   * 2. Undefined package ID set to "cloud"
+   * 3. Undefined version set to the latest file version
+   * 4. Check sil_euro_latin only added once
+   * Using the old keyboards list as paramter instead of keyboards file so this can be
+   * unit test.
+   * @param context Context
+   * @param dat_list ArrayList<HashMap<String, String>> Old keyboards list
+   * @param list List<Keyboard> Migrated keyboards list
+   */
+  public static void updateOldKeyboardsList(Context context, ArrayList<HashMap<String, String>> dat_list,
+                                            List<Keyboard> list) {
+    if (dat_list == null || list == null) {
+      return;
+    }
+
+    boolean defaultKeyboardInstalled = false;
+    for(HashMap<String, String> kbdMap : dat_list) {
+      boolean isNewKeyboard = kbdMap.containsKey(KeyboardPickerActivity.KMKEY_INTERNAL_NEW_KEYBOARD) &&
```

Comment:
Sorry if this question has been answered before elsewhere, but... what is this line about?  The "internal new keyboard" thing?  Why would it vary between "true" and "false" as part of the resource migration process?

Reply:
The "internal new keyboard" flag is used after installing a keyboard to highlight the keyboard in the keyboard list with a "[new]" string. When the keyboard gets switched, the flag is removed.

I can simplify things by just setting this to false during the keyboard list migration.